### PR TITLE
[MRG] Updates to network read and write

### PR DESCRIPTION
**Pull request by gtdang**
Working PR to update network HDF5 read and write.

### CODE REVIEW

File: hnn_core/tests/test_io.py [hunk 1]
```diff
@@ -1,26 +1,21 @@
...
                          [law_2021_model, calcium_model,
                           jones_2009_model])
 def test_network_io(tmp_path, network_model):
     # For simulation to be shorter(Discuss)
-    params = op.join(hnn_core_root, 'param', 'default.json')
-    if isinstance(params, str):
-        params = read_params(params)
+    params_path = Path(hnn_core_root, 'param', 'default.json')
+    params = read_params(params_path)
     params['N_pyr_x'] = 3
```
Comment (ntolley) [inline hnn_core/tests/test_io.py:19]: @gtdang the problem is here btw! At the moment we don't have a real API that let's us change the number of cells in the network, this is just a hack in the tests because smaller networks simulate much faster.

Essentially the saved network is a 3x3 grid, but then the loaded network is the default 10x10 grid.

One way this could be remedied is allowing users to pass `net  = Network(..., n_pyr_x=3, n_pyr_y=3)` explicitly so that the network size is saved in the HDF5 file.

@rythorpe @jasmainak any thoughts?

Comment (ntolley) [reply hnn_core/tests/test_io.py:19]: For context @rythorpe and @jasmainak, @gtdang found that the `net.connectivity` changed after writing to file and reloading (it's changing because the networks are of different sizes)

Comment (ntolley) [reply hnn_core/tests/test_io.py:19]: I believe the last point is the most accurate. All of the functions to setup the cell positions and grid size are called right when the network is initialized...

Comment (rythorpe) [reply hnn_core/tests/test_io.py:19]: @ntolley you're advocating for an API change here right? If so, I'd definitely support that. My vote would be for a single argument like `net = Network(..., mesh_shape=(3, 3))`. This is nitpicky, but my use of "mesh" instead of "grid" comes from the desire to support more flexible network cell positioning functions that e.g. place cells in hexogonal patterns or random patterns on a 2D plane rather than on a rectangular grid.

Comment (jasmainak) [reply hnn_core/tests/test_io.py:19]: +1 of making this explicit in our API. Does this apply only to pyramidal cells or basket cells as well? Do we need a dict to specify the cell type?

File: hnn_core/io.py [hunk 2]
```diff
@@ -37,54 +37,45 @@ def write_network(net, fname, overwrite=True, write_output=True):
...
+        'N_pyr_y': net._N_pyr_y,
+        'celsius': net._params['celsius'],
+        'cell_types': {name: template.to_dict()
+                       for name, template in net.cell_types.items()
+                       },
+        'gid_ranges': {cell: {'start': c_range.start, 'stop': c_range.stop}
+                       for cell, c_range in net.gid_ranges.items()
+                       },
+        'pos_dict': {cell: pos for cell, pos in net.pos_dict.items()},
+        'cell_response': _write_cell_response(net, write_output),
```
Comment (jasmainak) [inline hnn_core/io.py:67]: rename _write_cell_response? maybe _cell_response_to_dict?

File: hnn_core/io.py [hunk 3]
```diff
@@ -37,54 +37,45 @@ def write_network(net, fname, overwrite=True, write_output=True):
...
+        'external_drives': {drive: _external_drive_to_dict(params, write_output)
+                            for drive, params in net.external_drives.items()
+                            },
+        'external_biases': net.external_biases,
+        'connectivity': _write_connectivity(net.connectivity),
```
Comment (jasmainak) [inline hnn_core/io.py:72]: same comment ... something other than `_write_connectivity` ? because we don't write with the function. It's an auxiliary function

File: hnn_core/network.py [hunk 4]
```diff
@@ -432,9 +438,13 @@ def __eq__(self, other):
                                           target_gids=target_gids,
                                           loc=conn['loc'],
                                           receptor=conn['receptor'])
-            if len(match_conns) == 0:
+            if len(match_conns) == 1:
+                # check if connections are equal
+                if conn != other.connectivity[match_conns[0]]:
+                    return False
+            else:
```
Comment (gtdang) [inline hnn_core/network.py:445]: @ntolley I've made a change here to make sure it is comparing the connectivity contents between the to networks. Is there a possibility where the length of `match_conns` is greater than 1? Right now it will just return a False for those instances. Is that OK?

Comment (ntolley) [reply hnn_core/network.py:445]: It would be uncommon but it is totally possible to have two `_Connectivity` objects that are the same on all attributes searched by `pick_connection`.

It's not something that would frequently come up, but there's good reasons not to prevent people from calling `add_connection` multiple times

Comment (gtdang) [reply hnn_core/network.py:445]: Ah ok. So the current logic first checks that the new networks objects have equal number of connections. Once that clears it goes on to check the equivalency of connections using the pick_connection to find matches. Does the pick_connection match on all connection attributes or just a subset? In other words could c1 match with c2 but not actually be equivalent for all attributes?

File: hnn_core/tests/test_io.py (hunk 1 above)
Comment (ntolley) [reply hnn_core/tests/test_io.py:222]: Just discussed with @gtdang, we're thinking that an intermediate solution will be to implement `getter` and `setter` decorators for these variables.

I think making and API through the network `__init__()` will be worthwhile in the future, but we might want to discuss these details (like a specified mesh) in a separate PR

File: hnn_core/network.py [hunk 5]
```diff
@@ -438,10 +438,10 @@ def __eq__(self, other):
                                           target_gids=target_gids,
                                           loc=conn['loc'],
                                           receptor=conn['receptor'])
-            if len(match_conns) == 1:
-                # check if connections are equal
-                if conn != other.connectivity[match_conns[0]]:
+            if len(match_conns) >= 1:
+                if not any([conn == other.connectivity[match] for match in match_conns]):
```
Comment (gtdang) [inline hnn_core/network.py:442]: @ntolley Just realized the edge case we were trying to think of. 
```
net1.connectivity = [c1, c1, c2]
net2.connectivity = [c1, c2, c2]
```
The current logic wouldn't catch cases with this type of duplication since they are of equal lengths and each connection would find at least one match in the other set. Is this worth catching?

File: hnn_core/tests/test_io.py (hunk 1 above)
Comment (ntolley) [reply hnn_core/tests/test_io.py:249]: @gtdang actually I just realized this won't work because `net.connectivity` is populated when we call `jones_2009_model()`. So this would actually require having a `mesh_shape()` argument for all of the implemented models, and then pass that to `Network`.

It's not the prettiest solution but I'll try and give it a shot.

**Comment by gtdang** [comment]
@ntolley 
Got all the tests working on this PR. There's still one outstanding comment about an [edge case](https://github.com/jonescompneurolab/hnn-core/pull/704#discussion_r1465449050). Any thoughts on that? Other than that a final review of the test in io_test.py would be good before merge.

File: hnn_core/network.py (hunk 5 above)
Comment (ntolley) [reply hnn_core/network.py]: I don't have an immediate answer right now but this discussion provides some good solutions to the problem:
https://stackoverflow.com/questions/7828867/how-to-efficiently-compare-two-unordered-lists-not-sets?noredirect=1&lq=1

Comment (gtdang) [reply hnn_core/network.py]: Hmmm can test out if that counter method works! Thanks!

Comment (gtdang) [reply hnn_core/network.py]: I updated the eq function with the 3rd option from that thread. Now it will catch those edge cases.

**Comment by ntolley** [comment]
@gtdang great work!!! :tada: :tada:

File: hnn_core/tests/test_io.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+                   )
+    assert not any([bool(val['voltages'])
+                    for val in hdf5_read['rec_arrays'].values()]
+                   )
+    assert not any([bool(val['events'])
+                    for val in hdf5_read['external_drives'].values()]
+                   )
+
+
+def test_cell_response_to_dict(jones_2009_network):
```
Comment (jasmainak) [inline hnn_core/tests/test_io.py:177]: docstrings are missing in the test functions ... they are used by `pytest` when used in verbose mode. @gtdang can we add one-line docstrings to the test functions?

File: hnn_core/tests/test_io.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+hnn_core_root = Path(__file__).parents[1]
+assets_path = Path(hnn_core_root, 'tests', 'assets')
+
+
+@pytest.fixture
+def params():
+    params_path = Path(hnn_core_root, 'param', 'default.json')
+    params = read_params(params_path)
+    params['celsius'] = 37.0
+    params['threshold'] = 0.0
```
Comment (jasmainak) [inline hnn_core/tests/test_io.py:29]: why do you have to modify this? I'm guessing this doesn't get saved in the `hdf5` file?

Comment (gtdang) [reply hnn_core/tests/test_io.py:29]: I just copied specifications that were used when I refactored this script into smaller defined tests. There are some simulations being performed in the subsequent tests that are passed this fixture. I figured those parameters important for simulations?

Comment (jasmainak) [reply hnn_core/tests/test_io.py:29]: I see ... don't think we ever want to have a temperature other than 37 for humans ... my feeling is that we should hard-code it [here](https://github.com/jonescompneurolab/hnn-core/blob/df99e07272c8e3eba6b9f5647d547948ee7ae129/hnn_core/network_builder.py#L58) or alternatively add an argument to `simulate_dipole`. Doesn't seem to fit with the network object.

cc @rythorpe @ntolley
